### ISSUE TITLE
Meta information on blocks, items, and mobs.

### ISSUE THREAD

**Issue by TGlas** [Microsoft/malmo#340]
This is a feature request.

It would be nice if the Malmo library could provide more meta information about blocks, items, and mobs. For example, which block types block the agent's movement, which don't? Which ones block the view partially or completely? E.g., we can walk through a potato field, and we can see "through" the plant blocks. This is losely related to #173, however, the ability of blocks to drop as items is only one property of interest.

Simple example, the information whether a block type blocks movement or not is essential for path planning. Of course, this information can be learned from trial and error. But if I want to provide it to the agent as expert knowledge then it must be hard-coded, and doing this in the library is more efficient than everyone creating their own property lists (that's what I currently do, it is cumbersome and error-prone).

Is it possible to extend `Types.xsd` for this purpose? On the software level, an interface for querying properties by name would be nice, but compile time constants (arrays) are also fine.

Aside from technical considerations, it is not really clear which properties (formulated in high-level terms of game mechanics) are useful, but anyway, providing a basic set is probably a good start.

**Comment by DaveyBiggers** [comment]
I take your point about the efficiency of hard-coding this expert knowledge centrally, but I'm not convinced the core Malmo platform is the place to do this. One thing we'd really like to avoid is re-implementing Minecraft logic in Malmo. We'd like to keep the roles separate: Minecraft provides the world; the agent experiences the world; Malmo provides the interface to make that happen. As far as possible, the agent should be learning directly from Minecraft, not from Malmo itself.

Perhaps the grid observations, when we revamp them, could distinguish between solid and permeable blocks - but you'd still have issues with things like water or cobwebs, which are technically permeable, but which slow the agent, or move them around. Minecraft is a rich environment, where most simple rules have exceptions - doors obstruct you, but not if they are open... closed doors obstruct you, but not if they are linked to a pressure-plate... lava obstructs you, but not if you are impervious to fire... There's a limit to what properties we can usefully expose before we find ourselves having to implement all of Minecraft's game logic in our own library...

We'll have a think, but when it comes to specific hard-coded expert knowledge, I suspect it's better if everyone provides their own. It's readily available on the internet, on all the various Minecraft wiki sites, and my suspicion is that every researcher will have different requirements.

**Comment by TGlas** [comment]
I understand that you don't want this to be part of Malmö. Keeping it slim and simple is good for many reasons. Lists of properties are hard to maintain. And sure, learning is preferred over engineering. However, sometimes we may want to bring an agent up to speed without solving all lower level problems first, or we may want to create a "teacher" agent. Whether this makes sense or not depends on the research question.

Driving your argument to the extreme, the lists of block and item lists should be removed. The information could be learned. No, please don't do that :) All I want to say is that the amout of hard-coded information is somewhat arbitrary.

One example where trial-and-error based learning is actually hopeless is figuring out the crafting recepies. Also as a human player you have to look them up on the wiki, because the space of possible recepies is huge, something like `pow(number of block and item types, 9)`. So they have to be made available to the agent, in one way or the other. For this case Malmö already provides a sensible solution via the "craft" command, in combination with listing all block and item types. Similarly, a function a la `bool AgentHost::can_I_enter_this_block(...)` could be helpful.

For many properties, I agree that rules are necessarily fuzzy. Covering 95% is easy, the rest would be prone to over-engineering. So this is hard to represent in data. Crawling the wiki could indeed be a sensible solution for the easy 95%, thanks for the hint. I may well go for that at some point. Still, others may have the same problem...

**Comment by TGlas** [comment]
A brief update on this issue. In the meantime I have crawled the wiki and extracted some properties of blocks and items, most of which are non-essential but nice to have. The wiki is not intended to be machine readable, so I had to catch many corner cases and the data is probably not 100% correct. It is okay for my purposes right now.

Just as a note in reference to #173: The information in the "Drops" category is extremely diverse and hard to parse, therefore I decided to ignore it for now.

One useful bit of information is whether blocks are "transparent", i.e., whether they block the view entirely or not. I use this to emulate vision from the grid observations. It is of course simplified, but it comes close to what I expect a minimal vision module to do: return all visible blocks in the vicinity.

The wiki does not answer the question whether a block can be walked through or not, so I added this data manually. Currently it is only one bit per block type. I'll have to define more categories, like "character cannot enter the block (otherwise suffocates)", "character can pass block in all directions", "character cannot pass bottom face" (e.g., lili pad).

**Comment by Bloodknight** [comment]
Some of this falls right into what i consider fun.

Seems to me that as suggested, malmo should not *initially* hold any data about items, any items, at all, but provide instead some method of adding basic data, in the form of subsets. Given that we are using satans own database format (XML) it is relatively trivial to add to the system properties as discovered by the agent. I do have some trouble understanding how this might be achieved right now considering the weird way things are done.

As far as recipes are concerned, i'm assuming you are talking about crafting the items you need, the crafting system of the PC minecraft is deliberately and arbitrarily asinine, thats as polite as i'm ever going to get, there should be a lookup table of items an actor needs, and there should be a list of materials needed to create that item. You can layer this lookup system as needed, and add extra learning as needed, the in built chain of things you need to do to make each thing you need to make before you can make item x (enchanted book of feather falling level 4 for example) should already be available based on previous work anyway.  But if the actor wants an iron pick, he should throw 3 iron and a couple sticks in the crafting table, and malmo can intercept and work some maqic, because i can guarantee that unless you provide infinite stacks of sticks and piles or iron ingots, i find it unlikely that these recipes will be discovered. If theres a probability genius who can quick do the math of how many potential combinations of crafting can be created in a 3x3 grid with eve the exact composition (3 iron, 2 sticks).

TL;DR Malmo needs to add a function whereby an item can be identified by the agent, multiple levels of identification are needed from basic through to full disclosure.
Also, crafting is nasty, needs fixing, perhaps upgrading to console version crafting, anybody arouind here have MC sources.... 😈

**Comment by TGlas** [comment]
@Bloodknight, I have two comments.

First, we should distinguish between information you intend to provide to the agent and those learned through interaction. I don't think that Malmö should ever care about managing the latter. That data belongs to your controller. It cannot be Malmö's task to manage the internal state of your control program. Malmö should only provide an interface to interact with the game. It turns out that some aspects of the game as very complex. Therefore we need some additional data (see below). However, I am only talking about static/constant data.

Second, I consider the larger part (combinatorial complexity) of the crafting problem solved, check out the `SimpleCraftCommand`:
https://github.com/Microsoft/malmo/blob/master/Schemas/MissionHandlers.xsd#L912
Sure, we still need lists of crafting ingredients. This is an example of additional *static* data that should be provided by Malmö. Actually, it should be provided by Minecraft, since realistically, without the wiki the game cannot be played.
Combining this information by recursively traversing the more and more basic ingredients needed for crafting a complex item is an inherent aspect of the game complexity. Call it asinine, but the long planning horizon required to craft complex items is one of the fascinating aspects of playing the game (manually or automated).
So I may end up writing another helper script to extract crafting ingredients from the wiki. Actually, that involves parsing html, which I consider even more diabolic than the basic subset of XML.

In short, I still see a need for some additional information provided by Malmö, but it is not all that bad :)

**Comment by Bloodknight** [comment]
The chain of construction from basic to complex is fine, the asinine part is the exact shape pattern you must fill in in order to create... anything, even in terms of minecraft this exists in precisely 2 of 6? 5? versions, and that can be disabled in one, leaving the PC as the one and only version where arbitrary patterns are required to craft, rather than simple material provision and item request.

Static constant data is only as useful as the current version of minecraft, however what i meant by allowing or telling malmo to provide specific sets of data, would be in the form of differentiation between block types based purely on what tool you need for example. Another sample set would be a base tool set, so that the agent can have access to an axe for trees, pick for stone/rock/ore and shovel for sh...and....

The other point i made was simply that minecraft knows everything, so why in the heck malmo needs to know what minecraft knows i don't know, what it needs is the functionality to get that data from minecraft.

Extracting data from 3rd party sources when you have access to the primary resource has me scratching my head tbh (not only primary resource but also the tools to completely deobfuscate the source code if needed)

Just as a side comment, it appears you misunderstood most of my previous posts, it could be my writing style of the fact that researchers and game developers don't speak the same english. 
My main point being that malmo should have the ability to provide what information you or i need to use to give the agent what (tools, info, resource) we want the agent to have, seems to me on the whole we agree on most points here :)

**Comment by TGlas** [comment]
Thanks for clarifying, I see you point :)

**Comment by DaveyBiggers** [comment]
Just revisiting this old thread.

"Extracting data from 3rd party sources when you have access to the primary resource has me scratching my head tbh"

Granted, but writing code so that other people can write code so that they can extract information which is already freely available on the internet... well, that doesn't make much sense either. Particularly when it's static information which only needs to be retrieved once, and writing the Malmo code to retrieve it would be non-trivial and time-consuming.

I'm closing this thread as being outside of Malmo's scope.

Although, FWIW, there is some test code in Malmo which will let you dump out a list of recipes... just hit "U" while in-game, and a file called "recipe_dump.txt" will get spat out into your Minecraft folder...  it will look like this:
```
1xtile.blockLapis = 9xitem.dyePowder.blue(Lapis Lazuli)
1xtile.blockCoal = 9xitem.coal(Coal)
1xtile.hayBlock = 9xitem.wheat(Wheat)
1xitem.ingotGold = 9xitem.goldNugget(Gold Nugget)
1xitem.ingotIron = 9xitem.ironNugget(Iron Nugget)
1xitem.rabbitStew = 0xtile.air(Air), 1xitem.rabbitCooked(Cooked Rabbit), 0xtile.air(Air), 1xitem.carrots(Carrot), 1xitem.potatoBaked(Baked Potato), 1xtile.mushroom(Mushroom), 0xtile.air(Air), 1xitem.bowl(Bowl), 0xtile.air(Air)
1xitem.rabbitStew = 0xtile.air(Air), 1xitem.rabbitCooked(Cooked Rabbit), 0xtile.air(Air), 1xitem.carrots(Carrot), 1xitem.potatoBaked(Baked Potato), 1xtile.mushroom(Mushroom), 0xtile.air(Air), 1xitem.bowl(Bowl), 0xtile.air(Air)
1xtile.melon = 9xitem.melon(Melon)
```

**Comment by TGlas** [comment]
Thanks David. The U command sounds very useful - although adding the file to the repository would have done the job. Nice machine readable data :)

I don't fully agree with the "information which is already freely available on the internet" part, since the whole point of my request is to have that information in machine readable (straightforward to parse) format. The information must exist somewhere in Minecraft's logic (that's Bloodknight's point), however, I understand that it may exist in logic, not necessarily in tabular form, and it seems to be hard to extract and/or to provide. That's a pity, and I can live with having to learn the information through interaction.
